Error: Failed to copy section  when linking libclang

On LLVM commit `f83726e6` when testing [wild-proxy](https://github.com/mati865/wild-proxy) by running `PATH=~/Projects/wild-proxy:$PATH CC=wild-gcc CXX=wild-g++ cmake -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_BUILD_TYPE=Release -G Ninja -DCMAKE_EXPORT_COMPILE_COMMANDS=ON ../llvm` and then `ninja` I expected it to pass, but it failed when linking.

Then also reproduced it using more typical setup:
```
❯ PATH=~/Projects/wild:$PATH cmake -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_BUILD_TYPE=Release -G Ninja ../llvm

❯ PATH=~/Projects/wild:$PATH ninja libclang
...

Error: Failed to copy section ``.text._ZN12_GLOBAL__N_119ARMMachObjectWriter16recordRelocationEPN4llvm16MachObjectWriterERNS1_11MCAssemblerEPKNS1_10MCFragmentERKNS1_7MCFixupENS1_7MCValueERm`` from file lib/libLLVMARMDesc.a @ ARMMachObjectWriter.cpp.o (75777 (296/1))

Caused by:
    Direct relocation (R_X86_64_PC32) to dynamic symbol from non-writable section, but copy relocations are disabled. symbol `llvm::hexdigit(unsigned int, bool)::LUT` (8292 local=2107) in file #2304 (9/0) (lib/libclangAST.a @ ASTContext.cpp.o) (ADDRESS)
```

Sure, I thought it'd just clutter the issue, but here you go:
```
❯ PATH=~/Projects/wild:$PATH ninja libclang
[1/2] Linking CXX shared library lib/libclang.so.21.0.0git
FAILED: lib/libclang.so.21.0.0git
: && /usr/bin/c++ -fPIC -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -fno-strict-aliasing -O3 -DNDEBUG  -Wl,-z,defs -Wl,-z,nodelete -Wl,--color-diagnostics  -Wl,--gc-sections  -Wl,--version-script,/home/mateusz/Projects/llvm-project/clang/tools/libclang/libclang.map -shared -Wl,-soname,libclang.so.21.0git -o lib/libclang.so.21.0.0git tools/clang/tools/libclang/CMakeFiles/libclang.dir/BuildSystem.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndex.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexCXX.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexCodeCompletion.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexDiagnostic.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexHigh.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexInclusionStack.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexUSRs.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CIndexer.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXComment.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXCursor.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXExtractAPI.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXIndexDataConsumer.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXCompilationDatabase.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXLoadedDiagnostic.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXSourceLocation.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXStoredDiagnostic.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXString.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/CXType.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/Indexing.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/FatalErrorHandler.cpp.o tools/clang/tools/libclang/CMakeFiles/libclang.dir/Rewrite.cpp.o  -Wl,-rpath,"\$ORIGIN/../lib:"  lib/libclangAST.a  lib/libclangBasic.a  lib/libclangDriver.a  lib/libclangExtractAPI.a  lib/libclangFrontend.a  lib/libclangIndex.a  lib/libclangLex.a  lib/libclangRewrite.a  lib/libclangSema.a  lib/libclangSerialization.a  lib/libclangTooling.a  -ldl  lib/libLLVMAArch64CodeGen.a  lib/libLLVMAArch64AsmParser.a  lib/libLLVMAArch64Desc.a  lib/libLLVMAArch64Disassembler.a  lib/libLLVMAArch64Info.a  lib/libLLVMAArch64Utils.a  lib/libLLVMAMDGPUCodeGen.a  lib/libLLVMAMDGPUAsmParser.a  lib/libLLVMAMDGPUDesc.a  lib/libLLVMAMDGPUDisassembler.a  lib/libLLVMAMDGPUInfo.a  lib/libLLVMAMDGPUUtils.a  lib/libLLVMARMCodeGen.a  lib/libLLVMARMAsmParser.a  lib/libLLVMARMDesc.a  lib/libLLVMARMDisassembler.a  lib/libLLVMARMInfo.a  lib/libLLVMARMUtils.a  lib/libLLVMAVRCodeGen.a  lib/libLLVMAVRAsmParser.a  lib/libLLVMAVRDesc.a  lib/libLLVMAVRDisassembler.a  lib/libLLVMAVRInfo.a  lib/libLLVMBPFCodeGen.a  lib/libLLVMBPFAsmParser.a  lib/libLLVMBPFDesc.a  lib/libLLVMBPFDisassembler.a  lib/libLLVMBPFInfo.a  lib/libLLVMHexagonCodeGen.a  lib/libLLVMHexagonAsmParser.a  lib/libLLVMHexagonDesc.a  lib/libLLVMHexagonDisassembler.a  lib/libLLVMHexagonInfo.a  lib/libLLVMLanaiCodeGen.a  lib/libLLVMLanaiAsmParser.a  lib/libLLVMLanaiDesc.a  lib/libLLVMLanaiDisassembler.a  lib/libLLVMLanaiInfo.a  lib/libLLVMLoongArchCodeGen.a  lib/libLLVMLoongArchAsmParser.a  lib/libLLVMLoongArchDesc.a  lib/libLLVMLoongArchDisassembler.a  lib/libLLVMLoongArchInfo.a  lib/libLLVMMipsCodeGen.a  lib/libLLVMMipsAsmParser.a  lib/libLLVMMipsDesc.a  lib/libLLVMMipsDisassembler.a  lib/libLLVMMipsInfo.a  lib/libLLVMMSP430CodeGen.a  lib/libLLVMMSP430AsmParser.a  lib/libLLVMMSP430Desc.a  lib/libLLVMMSP430Disassembler.a  lib/libLLVMMSP430Info.a  lib/libLLVMNVPTXCodeGen.a  lib/libLLVMNVPTXDesc.a  lib/libLLVMNVPTXInfo.a  lib/libLLVMPowerPCCodeGen.a  lib/libLLVMPowerPCAsmParser.a  lib/libLLVMPowerPCDesc.a  lib/libLLVMPowerPCDisassembler.a  lib/libLLVMPowerPCInfo.a  lib/libLLVMRISCVCodeGen.a  lib/libLLVMRISCVAsmParser.a  lib/libLLVMRISCVDesc.a  lib/libLLVMRISCVDisassembler.a  lib/libLLVMRISCVInfo.a  lib/libLLVMSparcCodeGen.a  lib/libLLVMSparcAsmParser.a  lib/libLLVMSparcDesc.a  lib/libLLVMSparcDisassembler.a  lib/libLLVMSparcInfo.a  lib/libLLVMSPIRVCodeGen.a  lib/libLLVMSPIRVDesc.a  lib/libLLVMSPIRVInfo.a  lib/libLLVMSystemZCodeGen.a  lib/libLLVMSystemZAsmParser.a  lib/libLLVMSystemZDesc.a  lib/libLLVMSystemZDisassembler.a  lib/libLLVMSystemZInfo.a  lib/libLLVMVECodeGen.a  lib/libLLVMVEAsmParser.a  lib/libLLVMVEDesc.a  lib/libLLVMVEDisassembler.a  lib/libLLVMVEInfo.a  lib/libLLVMWebAssemblyCodeGen.a  lib/libLLVMWebAssemblyAsmParser.a  lib/libLLVMWebAssemblyDesc.a  lib/libLLVMWebAssemblyDisassembler.a  lib/libLLVMWebAssemblyInfo.a  lib/libLLVMWebAssemblyUtils.a  lib/libLLVMX86CodeGen.a  lib/libLLVMX86AsmParser.a  lib/libLLVMX86Desc.a  lib/libLLVMX86Disassembler.a  lib/libLLVMX86Info.a  lib/libLLVMXCoreCodeGen.a  lib/libLLVMXCoreDesc.a  lib/libLLVMXCoreDisassembler.a  lib/libLLVMXCoreInfo.a  lib/libLLVMCore.a  lib/libLLVMSupport.a  lib/libLLVMTargetParser.a  lib/libclangInstallAPI.a  lib/libLLVMTextAPIBinaryReader.a  lib/libclangFrontend.a  lib/libclangDriver.a  lib/libLLVMWindowsDriver.a  lib/libclangParse.a  lib/libclangSerialization.a  lib/libclangSema.a  lib/libclangAPINotes.a  lib/libclangEdit.a  lib/libclangAnalysis.a  lib/libclangSupport.a  lib/libLLVMFrontendHLSL.a  lib/libLLVMOption.a  lib/libclangFormat.a  lib/libclangToolingInclusions.a  lib/libclangToolingCore.a  lib/libclangRewrite.a  lib/libclangASTMatchers.a  lib/libclangAST.a  lib/libclangLex.a  lib/libclangBasic.a  lib/libLLVMAArch64Desc.a  lib/libLLVMAArch64Info.a  lib/libLLVMAArch64Utils.a  lib/libLLVMMIRParser.a  lib/libLLVMAMDGPUDesc.a  lib/libLLVMAMDGPUInfo.a  lib/libLLVMAMDGPUUtils.a  lib/libLLVMARMDesc.a  lib/libLLVMARMInfo.a  lib/libLLVMARMUtils.a  lib/libLLVMPasses.a  lib/libLLVMHipStdPar.a  lib/libLLVMCoroutines.a  lib/libLLVMHexagonDesc.a  lib/libLLVMHexagonInfo.a  lib/libLLVMLanaiDesc.a  lib/libLLVMLanaiInfo.a  lib/libLLVMLoongArchDesc.a  lib/libLLVMLoongArchInfo.a  lib/libLLVMipo.a  lib/libLLVMFrontendOpenMP.a  lib/libLLVMFrontendOffloading.a  lib/libLLVMObjectYAML.a  lib/libLLVMFrontendAtomic.a  lib/libLLVMVectorize.a  lib/libLLVMSandboxIR.a  lib/libLLVMLinker.a  lib/libLLVMRISCVDesc.a  lib/libLLVMRISCVInfo.a  lib/libLLVMSPIRVAnalysis.a  lib/libLLVMSystemZDesc.a  lib/libLLVMSystemZInfo.a  lib/libLLVMWebAssemblyDesc.a  lib/libLLVMWebAssemblyInfo.a  lib/libLLVMCFGuard.a  lib/libLLVMGlobalISel.a  lib/libLLVMIRPrinter.a  lib/libLLVMInstrumentation.a  lib/libLLVMAsmPrinter.a  lib/libLLVMSelectionDAG.a  lib/libLLVMCodeGen.a  lib/libLLVMScalarOpts.a  lib/libLLVMAggressiveInstCombine.a  lib/libLLVMInstCombine.a  lib/libLLVMCGData.a  lib/libLLVMBitWriter.a  lib/libLLVMObjCARCOpts.a  lib/libLLVMTarget.a  lib/libLLVMTransformUtils.a  lib/libLLVMAnalysis.a  lib/libLLVMProfileData.a  lib/libLLVMSymbolize.a  lib/libLLVMDebugInfoPDB.a  lib/libLLVMDebugInfoCodeView.a  lib/libLLVMDebugInfoMSF.a  lib/libLLVMDebugInfoBTF.a  lib/libLLVMDebugInfoDWARF.a  lib/libLLVMObject.a  lib/libLLVMMCParser.a  lib/libLLVMIRReader.a  lib/libLLVMBitReader.a  lib/libLLVMAsmParser.a  lib/libLLVMCore.a  lib/libLLVMRemarks.a  lib/libLLVMBitstreamReader.a  lib/libLLVMTextAPI.a  lib/libLLVMCodeGenTypes.a  lib/libLLVMMCDisassembler.a  lib/libLLVMMC.a  lib/libLLVMBinaryFormat.a  lib/libLLVMTargetParser.a  lib/libLLVMSupport.a  lib/libLLVMDemangle.a  -lrt  -ldl  -lm  /usr/lib/libz.so  /usr/lib/libzstd.so && :
WARNING: wild: --plugin /usr/lib/gcc/x86_64-pc-linux-gnu/14.2.1/liblto_plugin.so is not yet supported
WARNING: wild: -z defs is not yet supported
WARNING: wild: -z nodelete is not yet supported
Error: Failed to copy section ``.text._ZNK4llvm14NVPTXSubtarget25failIfClustersUnsupportedERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE`` from file lib/libLLVMNVPTXCodeGen.a @ NVPTXSubtarget.cpp.o (98560 (385/0))

Caused by:
    Direct relocation (R_X86_64_PC32) to dynamic symbol from non-writable section, but copy relocations are disabled. symbol `{vtable(llvm::support::detail::provider_format_adapter<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>)}` (100733 local=871) in file #20992 (82/0) (lib/libclangDriver.a @ Driver.cpp.o) (ADDRESS)

Stack backtrace:
   0: anyhow::error::<impl anyhow::Error>::msg
   1: libwild::layout::GroupState::do_pending_work
   2: rayon_core::broadcast::BroadcastContext::with
   3: <rayon_core::job::ArcJob<BODY> as rayon_core::job::Job>::execute
   4: rayon_core::registry::WorkerThread::wait_until_cold
   5: rayon_core::registry::ThreadBuilder::run
   6: std::sys::backtrace::__rust_begin_short_backtrace
   7: core::ops::function::FnOnce::call_once{{vtable.shim}}
   8: std::sys::pal::unix::thread::Thread::new::thread_start
   9: <unknown>
  10: <unknown>
collect2: error: ld returned 255 exit status
ninja: build stopped: subcommand failed.
```
Also, the symbol looks like this:
```
❯ readelf -W -t lib/libclangAST.a | bat -p
...
  [378] .rodata._ZZN4llvm8hexdigitEjbE3LUT
       PROGBITS        0000000000000000 021b40 000011 00   0   0 16
       [0000000000000202]: ALLOC, GROUP
...
```
This is probably related to this comment: https://github.com/davidlattimore/wild/blob/82f0a0122c2a534f0238a61b2f5d9e52baa226f6/libwild/src/layout.rs#L2597-L2598

--- a/libwild/src/symbol_db.rs
+++ b/libwild/src/symbol_db.rs
@@ -27,6 +27,7 @@ use crate::version_script::VersionScript;
 use ahash::HashMap;
 use ahash::RandomState;
 use anyhow::Context;
+use anyhow::Error;
 use anyhow::bail;
 use crossbeam_queue::SegQueue;
 use itertools::Itertools;
@@ -382,6 +383,29 @@ impl<'data> SymbolDb<'data> {
         symbol_id
     }
 
+    /// Reads the symbol visibility from the original object.
+    fn input_symbol_visibility(&self, symbol_id: SymbolId) -> Visibility {
+        let file_id = self.file_id_for_symbol(symbol_id);
+        match &self.groups[file_id.group()] {
+            Group::Prelude(_) => Visibility::Default,
+            Group::Objects(parsed_input_objects) => {
+                let obj = &parsed_input_objects[file_id.file()];
+                let local_index = symbol_id.to_input(obj.symbol_id_range);
+
+                let Ok(obj_symbol) = obj.object.symbol(local_index) else {
+                    return Visibility::Default;
+                };
+
+                match obj_symbol.st_visibility() {
+                    object::elf::STV_PROTECTED => Visibility::Protected,
+                    object::elf::STV_HIDDEN => Visibility::Hidden,
+                    _ => Visibility::Default,
+                }
+            }
+            Group::Epilogue(_) => Visibility::Default,
+        }
+    }
+
     /// Returns a struct that can be used to print debug information about the specified symbol.
     pub(crate) fn symbol_debug(&self, symbol_id: SymbolId) -> SymbolDebug<'_, 'data> {
         SymbolDebug {
@@ -625,44 +649,27 @@ pub(crate) fn resolve_alternative_symbol_definitions<'data>(
     let mut buckets = take(&mut symbol_db.buckets);
     let error_queue = SegQueue::new();
 
-    let replacements: Vec<Vec<(SymbolId, SymbolId)>> = buckets
+    let replacements: Vec<Vec<SymbolReplacement>> = buckets
         .par_iter_mut()
         .map(|bucket| {
             let mut local_replacements = Vec::new();
 
-            for (first, alternatives) in replace(
+            process_alternatives(
                 &mut bucket.alternative_definitions,
-                HashMap::with_hasher(RandomState::new()),
-            ) {
-                match select_symbol(symbol_db, first, &alternatives, resolved) {
-                    Ok(selected) => {
-                        local_replacements.push((first, selected));
-                        for alt in alternatives {
-                            local_replacements.push((alt, selected));
-                        }
-                    }
-                    Err(err) => {
-                        error_queue.push(err);
-                    }
-                }
-            }
+                &mut local_replacements,
+                &error_queue,
+                symbol_db,
+                resolved,
+            );
 
-            for (first, alternatives) in replace(
+            process_alternatives(
                 &mut bucket.alternative_versioned_definitions,
-                HashMap::with_hasher(RandomState::new()),
-            ) {
-                match select_symbol(symbol_db, first, &alternatives, resolved) {
-                    Ok(selected) => {
-                        local_replacements.push((first, selected));
-                        for alt in alternatives {
-                            local_replacements.push((alt, selected));
-                        }
-                    }
-                    Err(err) => {
-                        error_queue.push(err);
-                    }
-                }
-            }
+                &mut local_replacements,
+                &error_queue,
+                symbol_db,
+                resolved,
+            );
+
             local_replacements
         })
         .collect();
@@ -681,13 +688,81 @@ pub(crate) fn resolve_alternative_symbol_definitions<'data>(
         )));
     }
 
-    for (a, b) in replacements.iter().flat_map(|r| r.iter()) {
-        symbol_db.replace_definition(*a, *b);
+    for r in replacements.iter().flat_map(|r| r.iter()) {
+        // TODO: Currently we only make the symbol non-interposable, but we should also actually
+        // change its visibility too. We need somewhere to store this information. We also need
+        // linker-diff to report when we get exported dynamic symbols wrong.
+        if r.visibility != Visibility::Default {
+            symbol_db.symbol_value_flags[r.replace.as_usize()] |= ValueFlags::NON_INTERPOSABLE;
+        }
+        symbol_db.replace_definition(r.replace, r.selected);
     }
 
     symbol_db.buckets = buckets;
 
     Ok(())
+}
+
+#[derive(Clone, Copy)]
+struct SymbolReplacement {
+    /// The symbol to be replaced.
+    replace: SymbolId,
+
+    /// The symbol that was selected as the definition to use.
+    selected: SymbolId,
+
+    visibility: Visibility,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+enum Visibility {
+    Default,
+    Protected,
+    Hidden,
+}
+
+fn process_alternatives(
+    alternative_definitions: &mut HashMap<SymbolId, Vec<SymbolId>>,
+    local_replacements: &mut Vec<SymbolReplacement>,
+    error_queue: &SegQueue<Error>,
+    symbol_db: &SymbolDb,
+    resolved: &[ResolvedGroup],
+) {
+    for (first, alternatives) in replace(
+        alternative_definitions,
+        HashMap::with_hasher(RandomState::new()),
+    ) {
+        // Compute the most restrictive visibility of any of the alternative definitions. This is
+        // the visibility we'll use for our selected symbol. This seems like odd behaviour, but it
+        // matches what GNU ld appears to do and some programs will fail to link if we don't do
+        // this.
+        let visibility = alternatives
+            .iter()
+            .fold(symbol_db.input_symbol_visibility(first), |vis, id| {
+                vis.max(symbol_db.input_symbol_visibility(*id))
+            });
+
+        match select_symbol(symbol_db, first, &alternatives, resolved) {
+            Ok(selected) => {
+                local_replacements.push(SymbolReplacement {
+                    replace: first,
+                    selected,
+                    visibility,
+                });
+
+                for alt in alternatives {
+                    local_replacements.push(SymbolReplacement {
+                        replace: alt,
+                        selected,
+                        visibility,
+                    });
+                }
+            }
+            Err(err) => {
+                error_queue.push(err);
+            }
+        }
+    }
 }
 
 /// Selects which version of the symbol to use.

--- a/wild/tests/integration_tests.rs
+++ b/wild/tests/integration_tests.rs
@@ -2235,6 +2235,7 @@ fn integration_test(
         "rust-tls.rs",
         "input_does_not_exist.c",
         "ifunc2.c",
+        "visibility-merging.c",
         "tls-local-exec.c",
         "undefined_symbols.c",
         "whole_archive.c",

--- a/wild/tests/sources/visibility-merging-1.c
+++ b/wild/tests/sources/visibility-merging-1.c
@@ -1,0 +1,8 @@
+// Even though the definition of `data1` from the main file is given priority over this definition,
+// the fact that this definition is hidden causes the symbol to be hidden.
+int data1  __attribute__ ((weak, visibility(("hidden")))) = 0x100;
+
+// Similarly, this symbol makes the symbol of the same name from file protected.
+int data3  __attribute__ ((weak, visibility(("protected")))) = 0x55;
+
+int data4 __attribute__ ((weak, visibility(("hidden")))) = 0x99;

--- a/wild/tests/sources/visibility-merging.c
+++ b/wild/tests/sources/visibility-merging.c
@@ -1,0 +1,29 @@
+//#Static:false
+//#LinkArgs:-shared -z now
+//#Object:visibility-merging-1.c
+//#RunEnabled:false
+//#DiffIgnore:section.got
+
+// This symbol is included, but isn't exported as a dynamic symbol because of a second definition in
+// our other file that's marked as hidden.
+int data1 __attribute__ ((weak)) = 0x42;
+
+// This symbol is exported.
+int data2 __attribute__ ((weak)) = 0x88888888;
+
+// This symbol is exported, but is protected, since the definition in the second file is.
+int data3 __attribute__ ((weak)) = 0x55;
+
+// Protected here and hidden in our second object. Hidden should take priority.
+int data4 __attribute__ ((weak, visibility(("protected")))) = 0x99;
+
+// Make sure that direct references to `data1` work on account of it being hidden.
+int get_data1(void) {
+    return data1;
+}
+
+// Note, we don't check direct references to `data2` and `data3`. In the case of `data2`, direct
+// references wouldn't be permitted, since the symbol is interposable. In the case of `data3`, the
+// symbol will be protected, so direct references should be permitted, however GNU ld < 2.40 would
+// error in the case of direct references to protected symbols, so in order to allow our tests to
+// pass with such versions, we don't.

